File: apps/server-java/src/main/java/com/maisonamane/infrastructure/persistence/mongodb/mapper/PilotProductMapper.java
package com.maisonamane.infrastructure.persistence.mongodb.mapper;

import com.maisonamane.domain.pilot.aggregate.PilotProduct;
import com.maisonamane.domain.pilot.entity.ProductVariant;
import com.maisonamane.domain.pilot.enums.PriceRange;
import com.maisonamane.domain.pilot.enums.ProductCategory;
import com.maisonamane.domain.pilot.enums.ProductStatus;
import com.maisonamane.domain.pilot.enums.ProductType;
import com.maisonamane.domain.pilot.enums.Size;
import com.maisonamane.domain.pilot.enums.ViewType;
import com.maisonamane.domain.pilot.valueobject.*;
import com.maisonamane.infrastructure.persistence.mongodb.document.PilotProductDocument;
import com.maisonamane.infrastructure.persistence.mongodb.document.PilotProductDocument.*;
import java.util.stream.Collectors;

/**
 * Mapper between PilotProduct (domain) and PilotProductDocument (MongoDB).
 * Maps Effect-TS mapper functions to Java static methods.
 */
public final class PilotProductMapper {

    private PilotProductMapper() {
    }

    public static PilotProductDocument toDocument(PilotProduct product) {
        return new PilotProductDocument(
                product.id().value(),
                product.label().value(),
                product.type().name(),
                product.category().name(),
                product.description().value(),
                product.priceRange().name(),
                product.variants().stream().map(PilotProductMapper::variantToDocument).toList(),
                viewsToDocument(product.views()),
                product.status().name(),
                syncStatusToDocument(product.syncStatus()),
                product.createdAt(),
                product.updatedAt());
    }

    public static PilotProduct fromDocument(PilotProductDocument doc) {
        return new PilotProduct(
                ProductId.of(doc.id()),
                ProductLabel.of(doc.label()),
                ProductType.valueOf(doc.type()),
                ProductCategory.valueOf(doc.category()),
                ProductDescription.of(doc.description()),
                PriceRange.valueOf(doc.priceRange()),
                doc.variants().stream().map(PilotProductMapper::variantFromDocument).collect(Collectors.toList()),
                viewsFromDocument(doc.views()),
                ProductStatus.valueOf(doc.status()),
                syncStatusFromDocument(doc.syncStatus()),
                doc.createdAt(),
                doc.updatedAt());
    }

    // Variant mapping

    private static VariantDocument variantToDocument(ProductVariant variant) {
        return switch (variant) {
            case ProductVariant.StandardVariant std -> new VariantDocument(
                    "StandardVariant",
                    std.size().name(),
                    null,
                    null);
            case ProductVariant.CustomVariant custom -> new VariantDocument(
                    "CustomVariant",
                    "CUSTOM",
                    new DimensionsDocument(
                            custom.customDimensions().width().value(),
                            custom.customDimensions().length().value()),
                    custom.price().valueInCentimes());
        };
    }

    private static ProductVariant variantFromDocument(VariantDocument doc) {
        if ("CustomVariant".equals(doc.tag())) {
            return ProductVariant.custom(
                    CustomDimension.of(doc.customDimensions().width(), doc.customDimensions().length()),
                    Price.of(doc.price()));
        }
        return ProductVariant.standard(Size.valueOf(doc.size()));
    }

    // Views mapping

    private static ViewsDocument viewsToDocument(ProductViews views) {
        return new ViewsDocument(
                viewToDocument(views.front()),
                viewToDocument(views.detail()),
                views.additional().stream().map(PilotProductMapper::viewToDocument).toList());
    }

    private static ViewDocument viewToDocument(ProductView view) {
        return new ViewDocument(
                view.viewType().name(),
                view.imageUrl().value());
    }

    private static ProductViews viewsFromDocument(ViewsDocument doc) {
        return ProductViews.of(
                viewFromDocument(doc.front()),
                viewFromDocument(doc.detail()),
                doc.additional().stream().map(PilotProductMapper::viewFromDocument).toList());
    }

    private static ProductView viewFromDocument(ViewDocument doc) {
        return new ProductView(
                ViewType.valueOf(doc.viewType()),
                ImageUrl.of(doc.imageUrl()));
    }

    // SyncStatus mapping

    private static SyncStatusDocument syncStatusToDocument(SyncStatus status) {
        return switch (status) {
            case SyncStatus.NotSynced ignored -> new SyncStatusDocument(
                    "NotSynced", null, null, null, null, null);
            case SyncStatus.Synced synced -> new SyncStatusDocument(
                    "Synced",
                    synced.shopifyProductId().value(),
                    synced.syncedAt(),
                    null, null, null);
            case SyncStatus.SyncFailed failed -> new SyncStatusDocument(
                    "SyncFailed",
                    null, null,
                    new SyncErrorDocument(
                            failed.error().code(),
                            failed.error().message(),
                            failed.error().details()),
                    failed.failedAt(),
                    failed.attempts());
        };
    }

    private static SyncStatus syncStatusFromDocument(SyncStatusDocument doc) {
        return switch (doc.tag()) {
            case "Synced" -> SyncStatus.synced(
                    ShopifyProductId.of(doc.shopifyProductId()),
                    doc.syncedAt());
            case "SyncFailed" -> SyncStatus.syncFailed(
                    SyncError.of(
                            doc.error().code(),
                            doc.error().message(),
                            doc.error().details()),
                    doc.failedAt(),
                    doc.attempts());
            default -> SyncStatus.notSynced();
        };
    }
}

File: apps/server/src/infrastructure/persistence/mongodb/mappers/pilot-product.mapper.ts
// src/infrastructure/persistence/mongodb/mappers/pilot-product.mapper.ts

import {
  makeImageUrl,
  makePositiveCm,
  makePrice,
  makeProductDescription,
  makeProductId,
  makeProductLabel,
  makeShopifyProductId,
  type Material,
  type PilotProduct,
  type PredefinedSize,
  ProductShape,
  ProductStatus,
  ProductType,
  type ProductVariant,
  type SyncStatus,
  ViewType,
} from '../../../../domain/pilot'

// ============================================
// MONGODB DOCUMENT TYPE
// Uses string _id instead of ObjectId
// ============================================

interface PilotProductDocumentSchema {
  _id: string
  label: string
  type: string
  shape: string
  description: string
  material: string
  variants: Array<{
    sizeSpec:
      | { _tag: 'CatalogSize'; size: string }
      | { _tag: 'BespokeSize'; width: number; length: number }
    pricingSpec: { _tag: 'FormulaPrice' } | { _tag: 'NegotiatedPrice'; amount: number }
  }>
  views: {
    front: { viewType: string; imageUrl: string }
    detail: { viewType: string; imageUrl: string }
    additional: Array<{ viewType: string; imageUrl: string }>
  }
  status: string
  syncStatus: {
    _tag: 'NotSynced' | 'Synced' | 'SyncFailed'
    shopifyProductId?: string
    syncedAt?: Date
    error?: { code: string; message: string; details: unknown }
    failedAt?: Date
    attempts?: number
  }
  createdAt: Date
  updatedAt: Date
}

export type PilotProductDocument = PilotProductDocumentSchema

// ============================================
// MAPPER: Domain -> Document
// ============================================

export const pilotToDocument = (product: PilotProduct): PilotProductDocument => ({
  _id: product.id,
  label: product.label,
  type: product.type,
  shape: product.shape,
  description: product.description,
  material: product.material,
  variants: product.variants.map((v) => ({
    sizeSpec: v.sizeSpec,
    pricingSpec: v.pricingSpec,
  })),
  views: {
    front: {
      viewType: product.views.front.viewType,
      imageUrl: product.views.front.imageUrl,
    },
    detail: {
      viewType: product.views.detail.viewType,
      imageUrl: product.views.detail.imageUrl,
    },
    additional: product.views.additional.map((v) => ({
      viewType: v.viewType,
      imageUrl: v.imageUrl,
    })),
  },
  status: product.status,
  syncStatus: {
    _tag: product.syncStatus._tag,
    ...(product.syncStatus._tag === 'Synced'
      ? {
          shopifyProductId: product.syncStatus.shopifyProductId,
          syncedAt: product.syncStatus.syncedAt,
        }
      : {}),
    ...(product.syncStatus._tag === 'SyncFailed'
      ? {
          error: product.syncStatus.error,
          failedAt: product.syncStatus.failedAt,
          attempts: product.syncStatus.attempts,
        }
      : {}),
  } as PilotProductDocument['syncStatus'],
  createdAt: product.createdAt,
  updatedAt: product.updatedAt,
})

// ============================================
// MAPPER: Document -> Domain
// ============================================

export const pilotFromDocument = (doc: PilotProductDocument): PilotProduct => ({
  _tag: 'PilotProduct',
  id: makeProductId(doc._id),
  label: makeProductLabel(doc.label),
  type: doc.type as ProductType,
  shape: (doc.shape ?? ProductShape.STANDARD) as PilotProduct['shape'],
  description: makeProductDescription(doc.description),
  material: (doc.material ?? 'MTIRT') as Material,
  variants: doc.variants.map(mapVariant) as [ProductVariant, ...ProductVariant[]],
  views: {
    front: {
      viewType: doc.views.front.viewType as ViewType,
      imageUrl: makeImageUrl(doc.views.front.imageUrl),
    },
    detail: {
      viewType: doc.views.detail.viewType as ViewType,
      imageUrl: makeImageUrl(doc.views.detail.imageUrl),
    },
    additional: doc.views.additional.map((v) => ({
      viewType: v.viewType as ViewType,
      imageUrl: makeImageUrl(v.imageUrl),
    })),
  },
  status: doc.status as ProductStatus,
  syncStatus: mapSyncStatus(doc.syncStatus),
  createdAt: doc.createdAt,
  updatedAt: doc.updatedAt,
})

// ============================================
// HELPERS
// ============================================

const mapVariant = (v: PilotProductDocument['variants'][number]): ProductVariant => {
  const sizeSpec =
    v.sizeSpec._tag === 'BespokeSize'
      ? {
          _tag: 'BespokeSize' as const,
          width: makePositiveCm(v.sizeSpec.width),
          length: makePositiveCm(v.sizeSpec.length),
        }
      : {
          _tag: 'CatalogSize' as const,
          size: v.sizeSpec.size as PredefinedSize,
        }

  const pricingSpec =
    v.pricingSpec._tag === 'NegotiatedPrice'
      ? { _tag: 'NegotiatedPrice' as const, amount: makePrice(v.pricingSpec.amount) }
      : { _tag: 'FormulaPrice' as const }

  return { sizeSpec, pricingSpec }
}

const mapSyncStatus = (status: PilotProductDocument['syncStatus']): SyncStatus => {
  switch (status._tag) {
    case 'Synced':
      return {
        _tag: 'Synced',
        shopifyProductId: makeShopifyProductId(status.shopifyProductId!),
        syncedAt: status.syncedAt!,
      }
    case 'SyncFailed':
      return {
        _tag: 'SyncFailed',
        error: status.error!,
        failedAt: status.failedAt!,
        attempts: status.attempts!,
      }
    default:
      return { _tag: 'NotSynced' }
  }
}
